src/libcore/mem/maybe_uninit.rs

```diff
@@ -258,6 +258,37 @@ impl<T> MaybeUninit<T> {
         MaybeUninit { uninit: () }
     }
 
+    /// Create a new array of `MaybeUninit<T>` items, in an uninitialized state.
+    ///
+    /// # Examples
+    ///
```

Comment:
I think the example is complex enough to deserve an explanation. Maybe something like this:

> This example takes an immutable slice and transforms it element-wise to be processed further. As an optimization, if the slice is small enough, a stack allocation is used instead of a heap allocation. Unneeded initialization of the stack memory is avoided.

To emphasize the symmetry and contrast between `buffer` and `vec`, I'd consider changing their names to `stack` and `heap`, or maybe `array` and `vec`.

For further symmetry, maybe also change

```rust
buffer.iter_mut().zip(input).for_each(|(a, b)| { a.write(f(b)); });
```

to

```rust
input.iter().map(f).zip(buffer.iter_mut()).for_each(|a, b| { b.write(a); })
```

Reply:
Maybe you could use my `read` example instead (with a bit of cleanup)? It's much clearer what exactly is happening there.

Reply:
It is indeed, I’ve now used a tweaked version of it.